### PR TITLE
Feature/injectable regressor

### PR DESCRIPTION
**Pull request by mrkaiser**
You can now inject a regressor into a Lime explainer but if none is given a default of Ridge is used.

### CODE REVIEW

**Comment by marcotcr** [comment]
What's the use case for this? Do you find that you often want to use a regressor other than Ridge, or is it that you want to use Ridge with other parameters?

**Comment by mrkaiser** [comment]
Luckily it can enable both. This allows for users to experiment with different regressors other than Ridge, but if they do not want to change anything, it includes the default Ridge regressor as a the default regressor in Lime base. This allows for just a generally more robust library.

**Comment by marcotcr** [comment]
I understand it enables both. What I am asking is why we would want to enable this.
Every little thing can be parametrized - that does not make for a more robust library, it makes for one that is much harder to change. 
In this particular case, I have actually tried other regressors, but I was convinced that it didn't matter enough to worry about. I can be convinced otherwise.

**Comment by wdm0006** [comment]
From the perspective of including this kind of library in a production system, this PR would enable a few really useful things for me:

- performance/accuracy tradeoff.  If I'm using this to interpret something where throughput is really important, being able to tradeoff some accuracy (in the interpretation, determined by the score functionality previously merged in) by tuning tolerance or max_iter would potentially be useful.  
- consistency.  Being able to set the randomseed and the solver (currently it defaults to 'auto' which will change the solver depending on the input data), would make stable/idempotent testing and benchmarking easier.  

Of course trying out different regressors or regularization entirely is also interesting, as you mention, I wouldn't expect the differences to be all that significant. I think 90+% of users probably would leave the defaults, but similarly to the Ensemble methods in scikit-learn, being able to optionally inject this sort of things can be very useful for the software/testing aspect if not always the ML aspect.

**Comment by marcotcr** [comment]
Ok, I'm convinced. However, let's put the default value in lime_base, instead of adding it individually in LimeTabularExplainer and LimeTextExplainer. I'll add some comments to the code

File: lime/lime_base.py [hunk 1]
```diff
@@ -101,6 +101,7 @@ def explain_instance_with_data(self,
                                    distances,
                                    label,
                                    num_features,
+                                   model_regressor,
```
Comment (marcotcr) [inline lime/lime_base.py]: Let's put the default value here, make this an optional argument

File: lime/lime_tabular.py [hunk 2]
```diff
@@ -79,14 +83,15 @@ class LimeTabularExplainer(object):
     means and stds in the training data. For categorical features, perturb by
     sampling according to the training distribution, and making a binary feature
     that is 1 when the value is the same as the instance being explained."""
-    def __init__(self, training_data, feature_names=None, categorical_features=None,
+    def __init__(self, training_data, model_regressor=Ridge(alpha=1, fit_intercept=True), feature_names=None, categorical_features=None,
```
Comment (marcotcr) [inline lime/lime_tabular.py]: Let's make model_regressor an optional parameter which defaults to None.

File: lime/lime_tabular.py [hunk 3]
```diff
@@ -119,6 +124,11 @@ def __init__(self, training_data, feature_names=None, categorical_features=None,
             self.categorical_features = range(training_data.shape[1])
             discretized_training_data = self.discretizer.discretize(training_data)
 
+        if "sample_weight" not in inspect.signature(model_regressor.fit).parameters:
+            raise ValueError("the regressor's fit function must incorporate sample weights")
+        else:
+            self.model_regressor = model_regressor
+
```
Comment (marcotcr) [inline lime/lime_tabular.py]: These checks can be done in lime_base.py

File: lime/lime_text.py [hunk 4]
```diff
@@ -151,6 +156,7 @@ class LimeTextExplainer(object):
        Currently, we are using an exponential kernel on cosine distance, and
        restricting explanations to words that are present in documents."""
     def __init__(self,
+                 model_regressor=Ridge(alpha=1, fit_intercept=True),
```
Comment (marcotcr) [inline lime/lime_text.py]: Let's make model_regressor an optional parameter which defaults to None.

File: lime/lime_text.py [hunk 5]
```diff
@@ -178,6 +185,10 @@ def __init__(self,
                 word order in some way (bigrams, etc).
 
         """
+        if "sample_weight" not in inspect.signature(model_regressor.fit).parameters:
+            raise ValueError("the regressor's fit function must incorporate sample weights")
+        else:
+            self.model_regressor = model_regressor
```
Comment (marcotcr) [inline lime/lime_text.py]: let's move this check to lime_base.py

File: lime/tests/TestLimeTabular.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        feature_names = ['1', '2']
+        with self.assertRaises(ValueError):
+            explainer = LimeTabularExplainer(training_data=eye, class_names=class_names, model_regressor=lasso, feature_names=feature_names)
+
+    def test_lime_explainer_good_regressor(self):
+        class_names = ['foo', 'bar']
+        logit = LogisticRegression()
+        eye = np.eye(2, dtype=int)
+        feature_names = ['1', '2']
+        explainer = LimeTabularExplainer(training_data=eye, class_names=class_names, model_regressor=logit, feature_names=feature_names)
```
Comment (marcotcr) [inline lime/tests/TestLimeTabular.py]: I like the idea of having these tests, but now we will only need one for lime_base

File: lime/tests/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+__author__ = 'devgupta'
```
Comment (marcotcr) [inline lime/tests/__init__.py]: Let's not have this. I'm not writing author names on anything

**Comment by marcotcr** [comment]
Thanks!

**Comment by mrkaiser** [comment]
Cool, I'll make the change this weekend and push it. Thanks

**Comment by brunodoamaral** [comment]
This PR would allow me to use a more complex regressor, like XGBoost? It has a [feature_importances_](http://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.XGBClassifier.feature_importances_) which could be used with `method == 'highest_weights'`. @marcotcr does it make sense?

If it does, line [71 of lime_base.py](https://github.com/marcotcr/lime/blob/master/lime/lime_base.py#L71) should be changed to a more general function (instead of `clf.coef_`). Maybe the `model_regressor` parameter should be a tuple `(fit_function, score_function)`.

**Comment by brunodoamaral** [comment]
Well, I did some experiments trying to use a different model for explanation, but (at least) XGBoost is unable to return a equivalente of `clf.coef_`. Its `get_fscore()` function does not distinguish between the positive or negative contribution of each feature in the model (as does `clf.coef_`).

Therefore, it may be almost impossible to use it as a model explainer.

**Comment by mrkaiser** [comment]
@marcotcr  I pushed some updates, that solves your main concerns. Please let me know what you think. Thanks!

**Comment by mrkaiser** [comment]
Hey @marcotcr please let me know if this PR can continue or if there is anything I need to clean up before you can finish merging.

**Comment by marcotcr** [comment]
I merged it after removing python3 specific stuff (inspect) and making some minor changes.
